staging/src/k8s.io/apiserver/plugin/pkg/audit/log/backend.go

```diff
@@ -17,77 +17,195 @@ limitations under the License.
 	FormatLegacy = "legacy"
 	// FormatJson saves event in structured json format.
 	FormatJson = "json"
+
+	// TODO make this configurable
+	// worker numbers of routines that concurrently write events.
+	workers = 5
+	// Buffer up to 1000 events before blocking.
+	defaultBufferSize = 1000
 )
 
+// The plugin name reported in error metrics.
+const pluginName = "log"
+
 // AllowedFormats are the formats known by log backend.
 var AllowedFormats = []string{
 	FormatLegacy,
 	FormatJson,
 }
 
-type backend struct {
+type bufferBackend struct {
 	out          io.Writer
 	format       string
 	groupVersion schema.GroupVersion
+	// Channel to buffer events in memory before write them to file.
+	buffer chan *auditinternal.Event
 }
 
-var _ audit.Backend = &backend{}
+var _ audit.Backend = &bufferBackend{}
 
-func NewBackend(out io.Writer, format string, groupVersion schema.GroupVersion) audit.Backend {
-	return &backend{
+func NewBackend(out io.Writer, format, mode string, groupVersion schema.GroupVersion) (audit.Backend, error) {
+	switch mode {
+	case ModeBuffer:
+		return newBufferBackend(out, format, groupVersion), nil
+	case ModeDirect:
+		return newDirectBackend(out, format, groupVersion), nil
+	default:
+		return nil, fmt.Errorf("log mode %q is not in list of known modes (%s)",
+			mode, strings.Join(AllowedModes, ","))
+	}
+
+}
+
+func newBufferBackend(out io.Writer, format string, groupVersion schema.GroupVersion) audit.Backend {
+	return &bufferBackend{
 		out:          out,
 		format:       format,
 		groupVersion: groupVersion,
+		buffer:       make(chan *auditinternal.Event, defaultBufferSize),
 	}
 }
 
-func (b *backend) ProcessEvents(events ...*auditinternal.Event) {
+func (b *bufferBackend) ProcessEvents(events ...*auditinternal.Event) {
+	for i, event := range events {
+		// The following mechanism is in place to support the situation when audit
+		// events are still coming after the backend was shut down.
+		var sendErr error
+		func() {
+			// If the backend was shut down and the buffer channel was closed, an
+			// attempt to add an event to it will result in panic that we should
+			// recover from.
+			defer func() {
+				if err := recover(); err != nil {
+					sendErr = errors.New("audit log shut down")
+				}
+			}()
+
+			select {
+			case b.buffer <- event:
+			default:
+				sendErr = errors.New("audit log queue blocked")
+			}
+		}()
+		if sendErr != nil {
+			audit.HandlePluginError(pluginName, sendErr, events[i:]...)
+			return
+		}
+	}
+}
+
+func (b *bufferBackend) processEvent(ev *auditinternal.Event) {
+	logEvent(b.out, b.format, b.groupVersion, ev)
+}
+
+func (b *bufferBackend) worker() {
+	for event := range b.buffer {
+		b.processEvent(event)
+	}
+}
+
+func (b *bufferBackend) Run(stopCh <-chan struct{}) error {
+	glog.Infof("Starting audit log backend")
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(b.worker, time.Second, stopCh)
+	}
+
+	return nil
+}
+
+func (b *bufferBackend) Shutdown() {
+	close(b.buffer)
```

Comment:
Still need a WaitGroup to block on here

Reply:
I know the meaning. Need to process all audit events buffered before process exit, wright?

Reply:
Yup, basically block this call until all workers existed